lib/parsers/heif_parser.rb

```diff
@@ -0,0 +1,430 @@
+  }
+  HEIC_MIME_TYPE = 'image/heic'
+  HEIF_MIME_TYPE = 'image/heif'
+  # HEIC_SEQUENCE_MIME_TYPE = 'image/heic-sequence' # TODO: to be used when adding image-sequence parsing
+  # HEIF_SEQUENCE_MIME_TYPE = 'image/heif-sequence'
+
+  def self.call(io)
+    new.call(io)
+  end
+
+  def call(io)
+    @buf                = FormatParser::IOConstraint.new(io)
+    @format             = nil
+    @@major_brand       = nil
+    @width              = nil
+    @height             = nil
+    @exif_data_frames   = []
+    @compatible_brands  = nil
+    @metadata_start_pos = 0
+    @metadata_end_pos   = 0
+    @handler_type       = nil
+    @sub_items          = nil
+    @pixel_aspect_ratio = nil
+    @colour_info        = nil
+    @pixel_info         = nil
+    @horizontal_offset  = nil
+    @vertical_offset    = nil
+    @clean_aperture     = nil
+    @primary_item_id    = 0
+    @item_props         = {}
+    @rotation           = 0
+    @item_props_idxs    = []
+    @content_type       = nil
+    scan
+  end
+
+  def scan
+    # All HEIC files must be conform to ISO/IEC 23008-12:2017
+    # Moreover, all HEIC files are conform to ISO/IEC 14496-12:2015 and should be conform to the Clause 4 of such spec.
+    # Files are formed as a series of objects, called boxes. All data is contained in such boxes.
+    # All boxes start with a header which defines both size and type.
+    # The size is the entire size of the box, including the size and type header, fields, and all contained boxes.
+    # The fields in the objects are stored with the most significant byte first, commonly known as network byte order or big-endian format.
+    # A HEIC file must contain a File Type Box (ftyp).
+    # A file conforms to all the requirements of the brands listed in the compatible_brands.
+    scan_file_type_box
+
+    # file may be identified by MIME type of Annex C of ISO/IEC 23008-12 if 'mif1' is the major brand or Annex D if 'msf1' is the major brand.
+    # the MIME indicates the nature and format of our assortment of bytes
+    # note particularly that the brand 'mif1' doesn't mandate a MovieBox ("moov").
+    # One or more brands must be included in the list of compatible brands
+    return if @compatible_brands.nil?
+    if @compatible_brands&.include?(MIF1_MARKER)
+      scan_meta_level_box
+      if @major_brand == MIF1_MARKER
+        @content_type = HEIF_MIME_TYPE
+        @format = :heif
+      elsif (@compatible_brands & HEIC_MIME_POSSIBLE_TYPES.keys).length > 0
+        @format = :heic
+        @content_type = HEIC_MIME_TYPE
+      end
+    end
+    if @compatible_brands&.include?(MSF1_MARKER)
+      # TODO
+    end
+
+    result = FormatParser::Image.new(
+      format: @format,
+      width_px: @width,
+      height_px: @height,
+      intrinsics: {
+        'compatible_brands': @compatible_brands,
+        'handler_type': @handler_type,
+        # 'sub_items': @sub_items, # enable this if you want to output all the sub-itmes in the image
```

Comment:
Do we need this too?

Reply:
this can be really useful mainly for debugging reasons